### PR TITLE
autoload: refactor for conformance and performance

### PR DESCRIPTION
**Pull request by derekstavis**
Refactor autoload: Split the big function into two smaller ones,
doing only option parsing at main function.

The algorithm is also rewritten, now in two steps for both path
inclusion an exclusion functions: 1) use auxiliary lists to store
valid function and completion paths, 2) bulk insert or remove just
once in the variable.

Now also respects path insertion policy, keeping user function
path always in front of other paths, thus allowing precedence of
user functions.

### CODE REVIEW

File: lib/autoload.fish [hunk 1]
```diff
@@ -1,58 +1,53 @@
...
-    end
-  end
-
-  set -q return_success
-end
+      contains -- $path $fish_function_path
+        or set function_path $function_path $path
+    end;
+  end;
+  test "$function_path"
```
Comment (oranja) [inline lib/autoload.fish]: Wouldn't a more explicit `test -n` be more appropriate here?
(Same goes for line#34)

File: lib/autoload.fish [hunk 2]
```diff
@@ -1,58 +1,53 @@
...
-#   paths are missing, returns != 0.
-#
-#   When called with -e, the paths passed as arguments are removed from
-#   $fish_function_path. All arguments ending with `completions` are correctly
-#   removed from $fish_complete_path. Returns 0 if one or more paths erased. If
-#   no paths were erased, returns != 0.
-
-function autoload -d "Manipulate autoloading path components"
+function autoload
   set -l paths $argv
```
Comment (oranja) [inline lib/autoload.fish]: Not used anywhere anymore

Comment (derekstavis) [inline lib/autoload.fish]: It's used below!

File: lib/autoload.fish (hunk 1 above)
Comment (derekstavis) [inline lib/autoload.fish]: Yeah, it's also possible

File: lib/autoload.fish (hunk 2 above)
Comment (oranja) [inline lib/autoload.fish]: Just to be clear, I'm talking about `$paths`, and nope... :no_good:

Comment (derekstavis) [inline lib/autoload.fish]: Oh gosh, I was looking at the removed stuff :laughing:

**Comment by derekstavis** [comment]
Done! Rebased to current master!

**Comment by oranja** [comment]
:+1: :100:

**Pull request merged**